drivers: video: add imx219 sensor (RPi Cam v2)

josuah opened the pull request:
Split out of:

- https://github.com/zephyrproject-rtos/zephyr/pull/87004

Dependency:

- ~~https://github.com/zephyrproject-rtos/zephyr/pull/88003~~
- https://github.com/zephyrproject-rtos/zephyr/pull/88138
- https://github.com/zephyrproject-rtos/zephyr/pull/91983

This was only tested on hardware that is out-of-tree and at 1920x1080 resolution for now (which will be submitted upstream in maybe April):

- https://github.com/tinyvision-ai-inc/tinyvision_zephyr_sdk/blob/main/drivers/video/imx219.c

It works well though.
If you would rather have it submitted before https://github.com/zephyrproject-rtos/zephyr/pull/88003 is merged, let me know, I will refactor it so it can be merged earlier.

P.S.: Thank you [Alan Shaju](https://github.com/alanridel) for the original work adapted into this PR!

josuah commented:
Force-push:
- Rebase on main after #87872
- Added `Co-authored-by:` to the commit message

dleach02 commented:
@josuah please put this PR into draft mode until the dependent PRs merged.

andrewnyland commented:
Is there any way we can help with this? Looks great so far

josuah commented:
> Is there any way we can help with this? Looks great so far

It might be possible yes! I need to rebase this driver on the latest `main` (like we did [here](https://github.com/tinyvision-ai-inc/tinyvision_zephyr_sdk/blob/main/drivers/video/imx219.c)) and then test again on our own platform https://tinyclunx33.tinyvision.ai/

And then it will be possible to test it on any of the Zephyr platform that supports MIPI/CSI, which for now is mostly the i.MXRT1170 and friends, but more are coming!

I will have more time next week to prepare the way a little bit more...

josuah commented:
Short update on this: I will wait reception of i.MXRT1160 and STM32N6 kits, then remove the library from it, to make it more like the IMX335.

The approach of the "table-only" image sensor driver library was too rigid and does not fit how firmware/driver developers grasp image sensors. Better build tools that fit the community than make the community fit the tools! :)

josuah committed:
driver: video: ctrls: add VIDEO_CID_DIGITAL_GAIN

Only VIDEO_CID_ANALOGUE_GAIN and VIDEO_CID_GAIN were defined. Also add
the complementary VIDEO_CID_DIGITAL_GAIN.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
drivers: video: sort CMakeLists.txt/Kconfig imports

Switch to zephyr-keep-sorted-start/stop to keep checking the new
driver imports are always sorted, which reduces chances of git
conflict while adding new drivers from separate branches.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
drivers: video: introduce Sony IMX219 sensor driver

Add support for the Sony IMX219 CSI sensor.
This sensor supports resolution of 3280x2464 in RGGB bayer format
either 8 or 10 bits and using 2 or 4 CSI lanes.
Only 10 bits on 2 CSI lanes is currently supported, and only in
1920x1080 pixel resolution using cropping using the video_set_format(),
and will need to be converted to video_set_selection() instead.

Signed-off-by: Josuah Demangeon <me@josuah.net>
Co-authored-by: Alan Shaju <alanshaju@rideltech.com>
Co-authored-by: Alain Volmat <alain.volmat@foss.st.com>

josuah commented:
Taken using the USB3 UVC branch and running on the tinyCLUXN33 (Lattice CrossLinkU-NX33 FPGA module/devkit)

## Brightness (Black Level) control

![imx219_brightness_000](https://github.com/user-attachments/assets/accf0446-59e0-4cfd-b28b-749f11c0841a)
![imx219_brightness_500](https://github.com/user-attachments/assets/b29f43fe-c9f2-483e-bab5-002decd4fbf1)
![imx219_brightness_1023](https://github.com/user-attachments/assets/2333fbfc-e021-4b20-a3e0-31353d7a7ec9)

## Exposure control

![imx219_exposure_0000](https://github.com/user-attachments/assets/b1ef8a07-ff60-491f-952c-00d2278548fd)
![imx219_exposure_3000](https://github.com/user-attachments/assets/038a808a-af3c-46be-8d8b-51b78472c103)
![imx219_exposure_60000](https://github.com/user-attachments/assets/744189c4-20c9-41f3-9356-60f487d966b6)

## Gain control

![imx219_gain_100](https://github.com/user-attachments/assets/b391a4b4-3a60-4e2f-8d30-905fb71fcfd5)
![imx219_gain_1000](https://github.com/user-attachments/assets/e840fafb-4405-4098-8ff4-3ab6db4d5805)
![imx219_gain_2000](https://github.com/user-attachments/assets/1f10a99f-9647-4b7f-be02-9dd63e9532df)

## Test patterns

![imx219_test_pattern_color_bars](https://github.com/user-attachments/assets/5f0ee3a5-8a37-4d86-a28b-dfa9adf246b9)
![imx219_test_pattern_fade_to_grey](https://github.com/user-attachments/assets/34303296-1d27-421d-8478-a03f065fb5d3)
![imx219_test_pattern_pn9](https://github.com/user-attachments/assets/7b391b9e-1d3b-485c-99a8-4ee47fdf1e80)

Note: this is not tested using any in-tree hardware, but instead an external SDK on top of Zephyr native APIs... Though with no modification outside of the UVC and USB3 patches.

avolmat-st reviewed (changes requested):
Thanks for the driver. Looks pretty good. I've add some comments but nothing pretty serious, except maybe the pixel format which should be the PACKED version here I think for a CSI sensor. This could lead to issues in order to use it with exixting interfaces since translation from PACKED to UNPACKED is necessary, a thing done by the DCMIPP driver.

Could you also add the entry within the test/driver/build_all so that the driver can be test compiled ?

Which board have you been using ? Arducam one ? Would you be willing to add the shield for it as well ?
I am wondering if the reset GPIO might be mandatory to control or not, gpio handling is currently missing in this driver.

I have the RPi Cam v2 (imx219) board as well with with the proper adaptation flex cable (15 -> 22) as well so I could give it a try with the STM32N6 as well.

josuah commented on drivers/video/imx219.c:
```diff
@@ -0,0 +1,517 @@
+#define IMX219_REG_SOFTWARE_RESET	IMX219_REG8(0x0103)
+#define IMX219_REG_MODE_SELECT		IMX219_REG8(0x0100)
+#define IMX219_MODE_SELECT_STANDBY	0x00
+#define IMX219_MODE_SELECT_STREAMING	0x00
```
Definitely! I think I never called `video_stream_stop()` on it so failed to properly notice it.

josuah commented on drivers/video/imx219.c:
```diff
@@ -0,0 +1,517 @@
+
+/* Registers to crop down a resolution to a centered width and height */
+static const struct video_reg imx219_init_regs[] = {
+	{IMX219_REG_MODE_SELECT, 0x00},		/* Standby */
```
I forgot to modify it there too...

josuah commented on drivers/video/imx219.c:
```diff
@@ -0,0 +1,517 @@
+	{IMX219_REG_DPHY_CTRL, 0x00},		/* Timing auto */
+
+	/* Clock configuration registers */
+	{IMX219_REG_EXCK_FREQ, 24 << 8},	/* 24 MHz */
```
I was not sure whether it was better to expose all the sensor features out or just include what was tested.

josuah replied:
Done!

josuah commented on drivers/video/imx219.c:
```diff
@@ -0,0 +1,517 @@
+	{IMX219_REG_PLL_OP_MPY, 50},		/* Output clock multiplier */
+};
+
+static const struct video_reg imx219_size_1920x1080_regs[] = {
```
Maybe the format config should be the full frame size without cropping, and the cropping API coming on top.
I will wrap the other repeated fields into the existing macro `IMX219_REGS_CROP()` for now.

josuah commented on drivers/video/imx219.c:
```diff
@@ -0,0 +1,517 @@
+
+static const struct video_format_cap imx219_fmts[] = {
+	{
+		.pixelformat = VIDEO_PIX_FMT_SBGGR8,
```
I forgot to upgrade it since it was introduced into Zephyr... Good catch!

josuah replied:
For the `VIDEO_PIX_FMT_SBGGR8` in particular, there is no difference between the packed and non-packed version? However it is still required for the `VIDEO_PIX_FMT_SBGGR10P` output.

josuah commented on drivers/video/imx219.c:
```diff
@@ -0,0 +1,517 @@
+	}
+
+	switch (idx) {
+	case 0:
```
I was not sure what to do to avoid too many different implementations... so just went with `0` as a placeholder.

I'll go the same way the GC2145 goes: just switch on the width, ignoring the height... Sensors rarely have several *native* resolutions of same width anyway, that should work well...

https://github.com/tinyvision-ai-inc/zephyr/blob/main/drivers/video/gc2145.c#L940-L957

If you have suggestion I'd be glad to hear it.

josuah commented on drivers/video/imx219.c:
```diff
@@ -0,0 +1,517 @@
+		return ret;
+	}
+
+	return 0;
```
I'll use this convention now to reduce code size a bit. 👍

josuah committed:
tests: drivers: build_all: video: add IMX219

Build the IMX219 image sensor driver on the build-all tests.
No interconnection with any other device is done, which is not mandatory
at the time as no link-related parameter is used.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented:
Force-push:
- Address most change requests from @avolmat-st but a few are missing still
- Add the driver to the `build_all` CI tests.

josuah committed:
boards: shields: introduce the Raspberry Pi Camera Module 2

Add a shield for the Raspberry Pi Camera Module 2, featuring the IMX219
sensor. This module is suitable for bot the "normal" and "No IR" (no
infra-red) variants and uses the rasapberry pi connector.

Signed-off-by: Josuah Demangeon <me@josuah.net>
Co-authored-by: Alain Volmat <alain.volmat@foss.st.com>

josuah commented:
Removing NXP tag.
Removing reviewers auto-added due to original commits.
No vendors-specific or area-specific people to bother except for shields, doc, and video hardware.

avolmat-st reviewed (changes requested):
Haven't yet got time to test or review it properly yet but could you change the shield name within the index.rst and also it seems to me that at least PIXEL_RATE or LINK_FREQ control should be made available within the imx219 driver otherwise the CSI receiver won't be able to know the UI for the PHY.

kartben commented on drivers/video/CMakeLists.txt:
```diff
@@ -6,24 +6,26 @@ zephyr_library_sources(video_common.c)
+zephyr_library_sources_ifdef(CONFIG_VIDEO_OV7725 ov7725.c)
+zephyr_library_sources_ifdef(CONFIG_VIDEO_OV9655 ov9655.c)
 zephyr_library_sources_ifdef(CONFIG_VIDEO_SHELL video_shell.c)
-zephyr_library_sources_ifdef(CONFIG_VIDEO_SW_GENERATOR	video_sw_generator.c)
-zephyr_library_sources_ifdef(CONFIG_VIDEO_MT9M114	mt9m114.c)
-zephyr_library_sources_ifdef(CONFIG_VIDEO_OV7725	ov7725.c)
-zephyr_library_sources_ifdef(CONFIG_VIDEO_OV2640	ov2640.c)
-zephyr_library_sources_ifdef(CONFIG_VIDEO_GC2145	gc2145.c)
-zephyr_library_sources_ifdef(CONFIG_VIDEO_STM32_DCMI	video_stm32_dcmi.c)
-zephyr_library_sources_ifdef(CONFIG_VIDEO_OV5640	ov5640.c)
-zephyr_library_sources_ifdef(CONFIG_VIDEO_OV7670	ov7670.c)
-zephyr_library_sources_ifdef(CONFIG_VIDEO_OV9655	ov9655.c)
-zephyr_library_sources_ifdef(CONFIG_VIDEO_ESP32		video_esp32_dvp.c)
-zephyr_library_sources_ifdef(CONFIG_VIDEO_MCUX_SDMA	video_mcux_smartdma.c)
-zephyr_library_sources_ifdef(CONFIG_VIDEO_EMUL_IMAGER	video_emul_imager.c)
-zephyr_library_sources_ifdef(CONFIG_VIDEO_EMUL_RX	video_emul_rx.c)
-zephyr_library_sources_ifdef(CONFIG_VIDEO_IMX335	imx335.c)
-zephyr_library_sources_ifdef(CONFIG_VIDEO_ST_MIPID02	video_st_mipid02.c)
-zephyr_library_sources_ifdef(CONFIG_VIDEO_STM32_DCMIPP	video_stm32_dcmipp.c)
+zephyr_library_sources_ifdef(CONFIG_VIDEO_STM32_DCMI video_stm32_dcmi.c)
+zephyr_library_sources_ifdef(CONFIG_VIDEO_STM32_DCMIPP video_stm32_dcmipp.c)
+zephyr_library_sources_ifdef(CONFIG_VIDEO_ST_MIPID02 video_st_mipid02.c)
+zephyr_library_sources_ifdef(CONFIG_VIDEO_SW_GENERATOR video_sw_generator.c)
+# zephyr-keep-sorted-stop
```
where is imx219.c? :)

kartben commented on drivers/video/Kconfig:
```diff
@@ -58,42 +58,26 @@ config VIDEO_I2C_RETRY_NUM
 	  The default is to not retry. Board configuration files or user project can then
 	  use the number of retries that matches their situation.
 
+# zephyr-keep-sorted-start
```
where is Kconfig.imx219? :)

Commit:
shields: raspberry_pi_camera_module_2: add stm32n6570_dk configs files

Add board specific .conf files for the stm32n6570_dk.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

josuah commented:
Force-push:

- Rebase on top of latest Zephyr
- Fix bugs preventing it to be built and run [1]
- Set correct board name
- Integrate improvement from @avolmat-st branch
- Reduce the default exposure

[1]: I had to modify our SDK to support shields so I could test this thing end-to-end.

josuah commented:
I realize this is now dependent on this PR, so setting as draft in the meantime:
- https://github.com/zephyrproject-rtos/zephyr/pull/91983

josuah commented on tests/drivers/build_all/video/app.overlay:
```diff
@@ -132,6 +138,12 @@
+
+			test_i2c_imx219: imx219@10 {
+				compatible = "sony,imx219";
+				reg = <0x7>;
```
The `build_all` test does not actually run anywhere, and I2C addresses are just incrementing numbers just to avoid address clash.

The `@10` better be corrected to `@7` though, that is true!

kartben replied:
actually, test_i2c_mipid02 should be `@8`, test_i2c_ov9655 `@9`, and test_i2c_imx219 `@A`

josuah commented on drivers/video/imx219.c:
```diff
@@ -0,0 +1,576 @@
+		return -EINVAL;
+	}
+
+	ret = video_write_cci_reg(&cfg->i2c, IMX219_REG_MODE_SELECT, IMX219_MODE_SELECT_STANDBY);
```
This is a good point, better avoid too much "enable/disable" traffic.

josuah commented on drivers/video/imx219.c:
```diff
@@ -0,0 +1,576 @@
+		return ret;
+	}
+
+	ret = video_write_cci_reg(&cfg->i2c, IMX219_REG_MODE_SELECT, IMX219_MODE_SELECT_STREAMING);
```
This is a good point, better avoid too much "enable/disable" traffic.

josuah commented on drivers/video/imx219.c:
```diff
@@ -0,0 +1,576 @@
+		return ret;
+	}
+
+	fmt.width = imx219_fmts[0].width_min;
```
Yes, better get ahead of Coverity.

josuah commented on drivers/video/imx219.c:
```diff
@@ -0,0 +1,576 @@
+	/* Custom defaults */
+	{IMX219_REG_BINNING_MODE_H, 0x00},	/* No binning */
+	{IMX219_REG_BINNING_MODE_V, 0x00}	/* No binning */,
+	{IMX219_REG_DIGITAL_GAIN, 5000},
```
This makes sense, I will adjust.

josuah commented on drivers/video/imx219.c:
```diff
@@ -0,0 +1,576 @@
+	{IMX219_REG_BINNING_MODE_V, 0x00}	/* No binning */,
+	{IMX219_REG_DIGITAL_GAIN, 5000},
+	{IMX219_REG_ANALOG_GAIN, 240},
+	{IMX219_REG_INTEGRATION_TIME, 100},
```
This makes sense, I will adjust.

josuah commented on drivers/video/imx219.c:
```diff
@@ -0,0 +1,576 @@
+	{IMX219_REG_OPPXCK_DIV, 8},
+};
+
+static const struct video_reg imx219_fmt_raw10_regs[] = {
```
Advantage of using `SENSORNAME_CCI_` and `SENSORNAME_REG_` convention is that everything written before CCI integration (i.e. GC2145) is already compatible, and the naming convention can be applied as sensors are converted to CCI (if they need to).

josuah commented:
Force-push:

- Use `IMX219_REG_` prefix for macros encoding a direct address such as `0x1234`
- Use `IMX219_CCI_` prefix for macros encoding an address + flags such as `IMX219_REG8(0x1234)`

This way, CCI and non-CCI registers definitions can be interleaved in the driver without confusion of what is what.

- Apply some suggestions but not quite complete yet.
- Test with tinyCLUNX33 (tinyvision_zephyr_sdk) via west `patch apply` (needed for USB3 patch) but not yet STM32N6
- Rebase on top of latest main

josuah commented:
Roadmap for this:
- Move the cropping ability to the new `set_selection()` interface
- Rebase
- Re-test and re-check everything
